awx/main/models/jobs.py

```diff
@@ -869,31 +881,46 @@ def start_job_fact_cache(self, destination, modification_times, timeout=None):
                 with codecs.open(filepath, 'w', encoding='utf-8') as f:
                     os.chmod(f.name, 0o600)
                     json.dump(host.ansible_facts, f)
+                    log_data['written_ct'] += 1
+                    last_filepath_written = filepath
             except IOError:
                 system_tracking_logger.error('facts for host {} could not be cached'.format(smart_str(host.name)))
                 continue
-            # make note of the time we wrote the file so we can check if it changed later
-            modification_times[filepath] = os.path.getmtime(filepath)
-
-    def finish_job_fact_cache(self, destination, modification_times):
+        # make note of the time we wrote the last file so we can check if any file changed later
+        if last_filepath_written:
+            return os.path.getmtime(last_filepath_written)
```

Comment:
@jbradberry this is a scaled-back version of what I was trying to do in https://github.com/ansible/awx/pull/11641

We were passing this `modification_times` dict, which has a modtime for every host in the inventory with facts... so, ballpark that number at 35k keys... but really more (based on known case reports). Then that has to remain in memory for the duration of the job for comparison after it has finished.

Here, I replace the _dictionary_ with a _single value_. This fits with the goal of the rest of this PR - which is to addresses _scaling_ problems for facts with large inventories.

There's also no drawback I can see for this. We write all facts. Stop. We run the job. Stop. We read the facts. There's no way that we can _write the last fact_ before we _read the first fact_.

Reply:
I think that makes sense.